rules/framework/vfs_overlay.bzl

```diff
@@ -161,35 +180,76 @@ def _merge_vfs_infos(base, vfs_infos):
 
 # Internally the "vfs obj" is represented as a dictionary, which is keyed on
 # the name of the root. This is an opaque value to consumers
-def _make_vfs_info(roots):
+def _make_vfs_info(name, data):
     keys = {}
-    for root in roots:
-        name = root["name"]
-        keys[name] = root
+    keys[name] = data
     return keys
 
+# Roots must be computed _relative_ to the vfs_parent. It is no longer possible
+# to memoize VFS computations because of this.
+def _roots_from_datas(vfs_parent, datas):
+    roots = []
+    for data in datas:
+        roots.extend(_make_root(
+            vfs_parent = vfs_parent,
+            bin_dir_path = data.bin_dir_path,
+            build_file_path = data.build_file_path,
+            framework_name = data.framework_name,
+            root_dir = data.framework_path,
+            extra_search_paths = data.extra_search_paths,
+            module_map = data.module_map,
+            hdrs = data.hdrs,
+            private_hdrs = data.private_hdrs,
+            has_swift = data.has_swift,
+        ))
+    return roots
+
 def make_vfsoverlay(ctx, hdrs, module_map, private_hdrs, has_swift, merge_vfsoverlays = [], extra_search_paths = None, output = None):
     framework_name = ctx.attr.framework_name
     framework_path = "{search_path}/{framework_name}.framework".format(
         search_path = FRAMEWORK_SEARCH_PATH,
         framework_name = framework_name,
     )
 
-    roots = _make_root(ctx, framework_name, framework_path, extra_search_paths, module_map, hdrs, private_hdrs, has_swift)
-    vfs_info = _make_vfs_info(roots)
+    vfs_parent = _get_vfs_parent(ctx)
+
+    data = struct(
+        bin_dir_path = ctx.bin_dir.path,
+        build_file_path = ctx.build_file_path,
+        framework_name = framework_name,
+        framework_path = framework_path,
+        extra_search_paths = extra_search_paths,
+        module_map = module_map,
+        hdrs = hdrs,
+        private_hdrs = private_hdrs,
+        has_swift = has_swift,
+    )
+
+    roots = _make_root(
+        vfs_parent,
+        bin_dir_path = ctx.bin_dir.path,
+        build_file_path = ctx.build_file_path,
+        framework_name = framework_name,
+        root_dir = framework_path,
+        extra_search_paths = extra_search_paths,
+        module_map = module_map,
+        hdrs = hdrs,
+        private_hdrs = private_hdrs,
+        has_swift = has_swift,
+    )
+
+    vfs_info = _make_vfs_info(framework_name, data)
     if len(merge_vfsoverlays) > 0:
         vfs_info = _merge_vfs_infos(vfs_info, merge_vfsoverlays)
-        roots = vfs_info.values()
+        roots = _roots_from_datas(vfs_parent, vfs_info.values() + [data])
```

Comment:
`vfs_info.values() + [data]` is to merge the two arrays correct?

Reply:
vfs_info is simply a dict with key being framework_name and value being `data`, now we call `vfs_info.values()` which just returns an array whose only element is `data`, and then we merge it with `[data]?

Reply:
Yea exactly, you'll also find this in the factory's comment https://github.com/bazel-ios/rules_ios/pull/284/files#diff-450b1330485549712d2fd4e65b2b3fd0a7692fe8f4ec43a0b7b60bbb739313afL163

Reply:
Guess my confusion is that it is essentially merging two arrays whose elements are both just `data`? Am I correct? If so is this needed?

Reply:
The starlark dictionary function `.values` returns an array, so data is in an array add them together